### DISCUSSION TITLE
Anybody knows if Twitter API's work in Next.js?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#50963] [Help] [answered]
### Summary

More specific..  "twitter-api-v2" or "twitter-api-sdk"  ???

The people at Twitter are in full blown chaos right now, and nobody can get much tech information other than reading the docs, which of course doesn't say much...

The error I am getting would be this below:

```
./node_modules/twitter-api-v2/dist/esm/v1/client.v1.write.js
Module not found: Can't resolve 'fs' in 'C:\Users\Edgar\Desktop\Stuff\Acode\nextprojectd\node_modules\twitter-api-v2\dist\esm\v1'

Import trace for requested module:
./node_modules/twitter-api-v2/dist/esm/v1/client.v1.write.js
./node_modules/twitter-api-v2/dist/esm/v1/client.v1.js
./node_modules/twitter-api-v2/dist/esm/index.js
./pages/api/twitterClient.js

./node_modules/twitter-api-v2/dist/esm/v1/media-helpers.v1.js
Module not found: Can't resolve 'fs' in 'C:\Users\Edgar\Desktop\Stuff\Acode\nextprojectd\node_modules\twitter-api-v2\dist\esm\v1'
Import trace for requested module:
./node_modules/twitter-api-v2/dist/esm/v1/media-helpers.v1.js
./node_modules/twitter-api-v2/dist/esm/v1/client.v1.write.js
./node_modules/twitter-api-v2/dist/esm/v1/client.v1.js
./node_modules/twitter-api-v2/dist/esm/index.js
./pages/api/twitterClient.js


> Build failed because of webpack errors


```

**Comment by icyJoseph** [comment]
I think there's a couple of errors.

While I am not sure how the `twitter-api-v2` works, I can tell you that creating `pages/api/twitterClient` makes that an available route that others can reach. I guess you don't want that, rather you wanna use the client within a handler.

Second, you export from `pages/api/twitterClient`, into a page, https://github.com/Elindo586/nextprojectd/blob/ff1e8e37ad9b5a6e2bb034cb7f37bd46125f0335/pages/twitter/test.js#L8, and from the button handler, you invoke the twitter client. The issue here is that the `twitter-api-v2` package runs in Node.js only, and the click happens within the browser, where `fs` is not available.

So your path of action is to:

```js
// pseudocode
// pages/api/twitterClient

import { TwitterApi } from "twitter-api-v2";

const client = new TwitterApi({
  appKey: process.env.API_KEY,
  appSecret: process.env.API_SECRET,
  accessToken: process.env.ACCESS_TOKEN,
  accessSecret: process.env.ACCESS_SECRET,
});

// no idea if these are necessary btw
const bearer = new TwitterApi(process.env.BEARER_TOKEN);
const twitterClient = client.readWrite;
const twitterBearer = bearer.readOnly;

const workWithTheTwitterClient =  async(req, res) => {
    // here you receive data from your client for example:
   const body = req.body.message || "default tweet message" // or maybe, if not message, do not send anything at all
   await twitterClient.v2.tweet(body.message);

   return res.status(200).json({ok: 'ok'})
}

export default workWithTheTwitterClient;
```

Again, the `pages/api/route` are made to receive HTTP requests. They can come from anywhere, one of which can be your client.

So now in the client:

```jsx
// pseudo-code
      <button
        type="button"
        className="btn btn-primary"
        onClick={async () => {
          try {
            await fetch("/api/hello", {
              method: "POST",
              headers: {
                "content-type": "application/json",
              },
              body: JSON.stringify({ message: "Hello world" }),
            });
          } catch (e) {
            console.log(e); // fetch doesn't really throw on 4xx range
          }
        }}
      >
        Click
      </button>
```

**Reply by icyJoseph** [reply]
I found this starter kit, https://github.com/Dineshs91/nextjs-twitter-starter, while it uses another Twitter SDK, I think it is worth studying for your use case.

**Comment by Elindo586** [comment]
Hey thanks!   How have you been by the way?... always a pile of knowledge.

I am gonna have to try that in a bit when I get some time...

Now... I can run a simple  tweet ("twitter-api-v2")  post from my localhost without REACT or NEXT.JS... but I am not using any buttons... its so down to simple code that I have to do `node index.js` to run the code and send a tweet..   Is that why I didn't have to create any routes?  because I wasn't using the browser? 

You can see the sample test here:
https://github.com/Elindo586/treference/blob/main/index.js


Eventually I want to run it with a start/stop buttons which I would need the browser to run it, and put it in a cron job...

Now, I am gonna try to post on the twitter forums to find out if the twitter-api-v2 and/or twitter-api-sdk work in Next.js.... see if I get an answer back ...

thank you much!

**Reply by icyJoseph** [reply, marked as answer]
Well, they definitely do work with Next.js, - that's not the issue here. The problem is, how you implement the API. 

All you gotta do is send requests from the client to the API routes. That's done with the fetch API.

On the API routes you collect the inputs from the request and use them to command the Twitter API.

Where you have initially gone wrong is in trying to use their API, which is designed to - work in the Node.js environment, within a button click handler, which happens in a browser environment.
